core/src/main/scala/cilib/Archive.scala

```diff
@@ -0,0 +1,94 @@
+package cilib
+
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.numeric.Positive
+import scalaz.Scalaz._
+import scalaz._
+
+sealed trait ArchiveBound
+final case class Bounded[A](limit: Int Refined Positive, deletePolicy: List[A] => A) extends ArchiveBound
+final case class Unbounded() extends ArchiveBound
+
+sealed abstract class Archive[A] {
+  import Archive._
+
+  def values: List[A] =
+    this match {
+      case Empty(_, _)       => List()
+      case NonEmpty(l, _, _) => l
+    }
+
+  def bound: ArchiveBound =
+    this match {
+      case Empty(b, _)       => b
+      case NonEmpty(_, b, _) => b
+    }
+
+  def insertCondition: (A, A) => Boolean =
+    this match {
+      case Empty(_, c)       => c
+      case NonEmpty(_, _, c) => c
+    }
+
+  def insert(v: A): Archive[A] =
+    this match {
+      case Empty(b, c) => NonEmpty[A](List(v), b, c)
+      case NonEmpty(l, b, c) =>
+        b match {
+          case Bounded(limit, deletePolicy) =>
+            if (l.size < limit.value && l.forall(x => c(v, x)))
+              NonEmpty[A](v :: l, b, c)
+            else if (l.size == limit.value && l.forall(x => c(v, x))){
+              val selected = deletePolicy(l) // Intellij complains here? "Expected List[nothing] actual List[Any]"
```

Comment:
It could be possible that the deletion, or eviction, requires the removal of more than a single element. As a result, I don't think that this is a valid function at the moment.

Reply:
Hi @gpampara 

Does this mean that the 'deletePolicy: List[A] => A'  should be changed to 'deletePolicy: List[A] => List[A]'?